dot/sync/chain_sync.go

```diff
@@ -53,692 +49,691 @@ func (s chainSyncState) String() string {
-				Value:  peerset.BadBlockAnnouncementValue,
-				Reason: peerset.BadBlockAnnouncementReason,
-			}, p)
-			return fmt.Errorf("%w: for peer %s and block number %d",
-				errPeerOnInvalidFork, p, ps.number)
+		logger.Infof(
+			"🚣 currently syncing, %d peers connected, "+
+				"%d available workers, "+
+				"target block number %d, "+
+				"finalised block number %d with hash %s",
+			len(cs.network.Peers()),
+			cs.workerPool.totalWorkers(),
+			syncTarget, finalisedHeader.Number, finalisedHeader.Hash())
+
+		if isFarFromTarget {
+			cs.workerPool.useConnectedPeers()
+			err = cs.requestMaxBlocksFrom(bestBlockHeader)
+			if err != nil {
+				logger.Errorf("requesting max blocks from best block header: %s", err)
+			}
+		} else {
+			// we are less than 128 blocks behind the target we can use tip sync
+			cs.syncMode.Store(tip)
+			isSyncedGauge.Set(1)
+			logger.Debugf("switched sync mode to %d", tip.String())
+			return
 		}
+	}
+}
 
-		// peer is on a fork, check if we have processed the fork already or not
-		// ie. is their block written to our db?
-		has, err := cs.blockState.HasHeader(ps.hash)
-		if err != nil {
-			return fmt.Errorf("has header: %w", err)
-		}
+func (cs *chainSync) getSyncMode() chainSyncState {
+	return cs.syncMode.Load().(chainSyncState)
+}
 
-		// if so, do nothing, as we already have their fork
-		if has {
-			return nil
-		}
+// setPeerHead sets a peer's best known block
+func (cs *chainSync) setPeerHead(who peer.ID, bestHash common.Hash, bestNumber uint) {
+	cs.workerPool.fromBlockAnnounce(who)
+
+	cs.peerViewLock.Lock()
+	defer cs.peerViewLock.Unlock()
+
+	cs.peerView[who] = &peerView{
+		who:    who,
+		hash:   bestHash,
+		number: bestNumber,
+	}
+}
+
+func (cs *chainSync) onBlockAnnounce(announced announcedBlock) error {
+	if cs.pendingBlocks.hasBlock(announced.header.Hash()) {
+		return fmt.Errorf("%w: block %s (#%d)",
+			errAlreadyInDisjointSet, announced.header.Hash(), announced.header.Number)
+	}
+
+	err := cs.pendingBlocks.addHeader(announced.header)
+	if err != nil {
+		return fmt.Errorf("while adding pending block header: %w", err)
+	}
+
+	if cs.getSyncMode() != tip {
+		return nil
+	}
+
+	_, _, isFarFromTarget, err := cs.isFarFromTarget()
+	if err != nil && !errors.Is(err, errNoPeerViews) {
+		return fmt.Errorf("checking target distance: %w", err)
 	}
 
-	// the peer has a higher best block than us, or they are on some fork we are not aware of
-	// add it to the disjoint block set
-	if err = cs.pendingBlocks.addHashAndNumber(ps.hash, ps.number); err != nil {
-		return fmt.Errorf("add hash and number: %w", err)
+	if !isFarFromTarget {
+		return cs.requestAnnouncedBlock(announced)
 	}
 
-	cs.workQueue <- ps
-	logger.Debugf("set peer %s head with block number %d and hash %s", p, number, hash)
+	// we are more than 128 blocks behind the head, switch to bootstrap
+	cs.syncMode.Store(bootstrap)
+	isSyncedGauge.Set(0)
+	logger.Debugf("switched sync mode to %d", bootstrap.String())
+	go cs.bootstrapSync()
```

Comment:
No, this is inside the `onBlockAnnounce` method, so every time we receive a block announce from the network, among other things, we check if the current sync mode is `bootstrap`, if it is then we just return (which means that the `bootstrapSync` goroutine is running yet, when it finishes then it changes the current sync mode to `tip`) otherwise we check how far we are from the tip of the chain, if needed we execute the `bootstrapSync` goroutine again but before when change the current mode to `bootstrap` to avoid spawning the goroutine more than once